src/index.js

```diff
@@ -1,56 +1,246 @@
+        }, {
+          prefix: new Key('/'),
+          datastore: new LevelStore(path.join(this.path, levelDirectory), {
+            db: this.options.level
+          })
+        }])
 
-    const blobStores = initializeBlobStores(options.stores)
+        this.blockstore = blockstore(this)
+        this.closed = false
+        cb()
+      }
+    ], (err) => {
+      if (err && this.lockfile) {
+        return this.lockfile.close((err2) => {
+          log('error removing lock', err2)
+          callback(err)
+        })
+      }
+
+      callback(err)
+    })
+  }
 
-    const setup = (name, needs) => {
-      needs = needs || {}
-      const args = [repoPath, blobStores[name]]
-      if (needs.locks) {
-        args.push(this.locks)
+  /**
+   * Check if the repo is already initialized.
+   *
+   * @private
+   * @param {function(Error)} callback
+   * @returns {void}
+   */
+  _isInitialized (callback) {
+    parallel([
+      (cb) => this.config.exists(cb),
+      (cb) => this.version.check(repoVersion, cb)
+    ], (err, res) => {
+      if (err) {
+        return callback(err)
       }
 
-      if (needs.config) {
-        args.push(this.config)
+      if (!res[0]) {
+        return callback(new Error('repo is not initialized yet'))
       }
+      callback()
+    })
+  }
 
-      return stores[name].setUp.apply(stores[name], args)
+  /**
+   * Close the repo and cleanup.
+   *
+   * @param {function(Error)} callback
+   * @returns {void}
+   */
+  close (callback) {
+    if (this.closed) {
+      return callback(new Error('repo is already closed'))
     }
 
-    this.locks = setup('locks')
-    this.version = setup('version', {locks: true})
-    this.config = setup('config', {locks: true})
-    this.keys = setup('keys', {locks: true, config: true})
-    this.blockstore = setup('blockstore', {locks: true})
+    log('closing at: %s', this.path)
+    series([
+      (cb) => this._fsStore.delete(apiFile, (err) => {
+        if (err && err.message.startsWith('ENOENT')) {
+          return cb()
+        }
+        cb(err)
+      }),
+      (cb) => this.store.close(cb),
+      (cb) => this._fsStore.close(cb),
+      (cb) => {
+        this.closed = true
+        this.lockfile.close(cb)
+      }
+    ], callback)
   }
 
+  /**
+   * Check if a repo exists.
+   *
+   * @param {function(Error, bool)} callback
+   * @returns {void}
+   */
   exists (callback) {
     this.version.exists(callback)
   }
-}
 
-function initializeBlobStores (store) {
-  if (store.constructor) {
-    return {
-      keys: store,
-      config: store,
-      blockstore: store,
-      logs: store,
-      locks: store,
-      version: store
-    }
+  /**
+   * Get the private key from the config.
+   *
+   * @param {function(Error, string)} callback
+   * @returns {void}
+   */
+  getPrivateKey (callback) {
+    this.config.get((err, config) => {
+      if (err) {
+        return callback(err)
+      }
+      callback(null, config.Identity.PrivKey)
+    })
+  }
```

Comment:
Why do we need this method here? Repo should not assume a specific format of the config file.